docs: user + design docs for --json-schema structured output (#4051)
QwenLM/qwen-code: docs/design/structured-output/structured-output.md

free-form text.

## Approach: synthetic tool whose parameter schema IS the user schema

When `--json-schema` is set, `Config.createToolRegistry` registers a
synthetic `structured_output` tool
([`syntheticOutput.ts`](../../../packages/core/src/tools/syntheticOutput.ts)).
Its `parametersJsonSchema` is exactly the schema the user passed; its
`execute()` returns a stop-message `llmContent`. The function-calling
API already validates tool args against `parametersJsonSchema`, so
"the model returned an answer conforming to the schema" reduces to
"the model successfully called `structured_output`."

Three properties fall out of this for free:

1. **No bespoke validator path.** Ajv-backed `validateToolParams`
   already runs inside `BaseDeclarativeTool.build()` and rejects
   non-conforming args before `execute()` ever fires.
2. **Standard retry behavior.** A validation failure surfaces to the
   model as a tool-call error the same way any other tool's args error
   does. The model sees the Ajv message and can correct in the next
   turn.
3. **Provider-agnostic.** Gemini, OpenAI, and Anthropic all serialize
   tool param schemas the same way; the synthetic tool plugs into all
   three.

> The tool is registered with `alwaysLoad: true` so the ToolSearch

Author: `ToolSearch` is undefined for readers outside the codebase — a one-sentence gloss or link to its design doc (if one exists) would unblock the `alwaysLoad: true` rationale below. Right now "the ToolSearch on-demand-loading infrastructure" is opaque unless you already know what #3589 introduced.

Reviewer: Addressed in b0fe3a7. Added a parenthetical gloss: "the ToolSearch on-demand-loading infrastructure (introduced in #3589 — keeps the exposed tool surface small by deferring rarely-used tools behind a search call, only mounting their full schemas when the model asks)".

   [`SchemaValidator.compileStrict`](../../../packages/core/src/utils/schemaValidator.ts).
   A dedicated Ajv instance with `strictSchema: true` surfaces
   typos like `propertees` that the lenient runtime validator would
   silently swallow.

### `schemaRootAcceptsObject` boundaries

The walk is intentionally best-effort. It catches the unambiguous
"this can never accept an object" cases, and defers anything that
needs whole-schema satisfiability analysis to Ajv at runtime.

**Decided at parse time:**

| Pattern | Outcome |
| --- | --- |
| `type` present, doesn't include `"object"` | reject |
| `type: ["object", "null"]` etc. | accept |
| `const`: non-object value | reject |
| `enum`: no object members (incl. empty) | reject |
| `anyOf`/`oneOf`: empty array | reject |
| `anyOf`/`oneOf`: no branch admits object | reject |
| `allOf`: any branch is `false` or rejects object | reject |
| Root `$ref` (with or without sibling `type`) | reject |
| `not`: bare `{type: "object"[…]}` (no narrowing keywords) | reject |
| `not`: `{type: "object", required: […], …}` etc. | accept (narrowing keywords leave some objects satisfiable; defer) |
| `if: true` + `then` rejects object | reject |

> | `if: false` + `else` rejects object | reject |

Author: **[Suggestion]** The `{type: "object"[…]} ` syntax in the `not` row is ambiguous — it reads like JavaScript array indexing. The next row correctly uses `{type: "object", required: […], …}` with the ellipsis after the comma. Use the same pattern for consistency.

```suggestion
| `not`: bare `{type: "object"}` (no narrowing keywords) | reject |
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Reviewer: Addressed in b0fe3a7. The `not` row now reads:

```
| `not`: bare `{type: "object"}` (no narrowing keywords) | reject |
```

Consistent ellipsis style with the other rows.

- **Success path** (a structured call captured the contract this turn):
  `"Skipped: this turn's structured_output contract took precedence as
  the terminal output."` — short, because the session terminates
  immediately and no consumer (model or SDK) acts on it.
- **Retry path** (no structured call captured, the model gets another
  turn): adds `"Re-issue this call in a separate turn if needed."` —
  this is the only model-actionable case.

### Main-turn / drain-turn parity

`processToolCallBatch(batchRequests, setModelOverride)` is defined
inside `runNonInteractive` and called from both:

- The main-turn loop (top of the function).
- `drainOneItem` (cron-prompt / background-task notification reply
  loop).

The drain turn matters because `structured_output` is registered for
the whole session, so a cron job or a notification reply MIGHT also
fire the tool. The helper handles both call sites identically;
the only call-site-specific binding is which `modelOverride` variable
to write to — passed in as a setter.

### Structured success terminal block

`emitStructuredSuccess()` (also defined inside `runNonInteractive`) is
the shared "we got a valid call, shut down" path:

> 1. `registry.abortAll()` aborts in-flight background agents — the

Author: **[Suggestion]** The 500ms holdback is described as a reliable solution for orphaned events, but it's best-effort — if an agent's abort handler takes longer than 500ms, `task_started` events can still appear without matching `task_notification`. Consider adding a caveat.

```suggestion
Bounded holdback (`STRUCTURED_SHUTDOWN_HOLDBACK_MS = 500` ms) so the natural cancel handlers of just-aborted agents have a chance to emit their terminal `task_notification` and land it in `localQueue`. This is best-effort — orphaned `task_started` events remain possible under load if abort handlers exceed the timeout. Without the holdback, stream-json consumers would see `task_started` events without matching `task_notification`.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Reviewer: Addressed in cc830f6. The design doc now includes the best-effort caveat:

> "The 500 ms ceiling is best-effort — orphaned `task_started` events remain possible under load if a particular agent's abort handler exceeds the budget."

Also documents that the loop does **not** poll the abort signal, so a SIGINT during holdback won't short-circuit the result already captured.

[`BaseJsonOutputAdapter.buildResultMessage`](../../../packages/cli/src/nonInteractive/io/BaseJsonOutputAdapter.ts)
treats the presence of `structuredResult` (tracked via `'structuredResult' in options`,
not `!== undefined`, so the contract is preserved even when the model
called `structured_output` with no args under an empty schema):

- `result` is forced to `JSON.stringify(payload)` — overriding any
  free-text summary the adapter accumulated.
- A top-level `structured_result` field carries the raw object for
  consumers that don't want to re-parse the stringified form.
- `undefined` payloads normalize to `null` (rendered as the literal
  JSON `null` in both fields) so the field can't silently disappear.

TEXT mode writes just the `result` field + newline to stdout. JSON
mode batches the full event log. Stream-json mode emits each event on
its own line.

## Privacy: cross-surface redaction

The args submitted via `structured_output` ARE the structured payload,
already on stdout. Persisting them again on durable on-device surfaces
(or worse, exporting them off-device through telemetry) is duplication
that leaks the payload into longer-lived storage than the user asked

> for.

Author: The rationale here covers the success path ("args ARE the payload, already on stdout"). But validation-failure retries also get redacted (and the user doc correctly notes this), and those args may never reach stdout. The redaction is right either way — but the intent stated here would be cleaner as "never persist any args from this synthetic tool, regardless of outcome," rather than just dedup-vs-stdout.

Reviewer: Addressed in b0fe3a7. Reworded the intent to: "never persist any args from this synthetic tool, regardless of outcome" — covering both the success-path dedup rationale and the validation-failure case where args may never reach stdout.

Two surfaces have to redact, and both share the same placeholder
constant
[`STRUCTURED_OUTPUT_REDACTED_ARGS`](../../../packages/core/src/tools/syntheticOutput.ts):

- `ToolCallEvent.function_args` (telemetry) — covers OTLP exports,
  QwenLogger, ui-telemetry, and the chat-recording UI event mirror.
- `redactStructuredOutputArgsForRecording` (used by
  `recordAssistantTurn` in `geminiChat.ts`) — covers the on-disk
  chat-recording JSONL at `<projectDir>/chats/<sessionId>.jsonl`.
  Validation-failure retries land here too — each retry's args also
  get the same placeholder.

The shared constant prevents drift between the two surfaces. Tool-call
metrics (duration, success, decision) are preserved.

## Permission gating

> `structured_output` is in the deliberately-excluded set in

Author: Two issues with this phrasing:

1. "in the deliberately-excluded set in `PermissionManager.CORE_TOOLS`" reads as "in the excluded-set *inside* CORE_TOOLS" — but `CORE_TOOLS` IS the allowlist-subject set; synthetic tools are simply not members. Suggest: "`structured_output` is deliberately excluded from `PermissionManager.CORE_TOOLS` (the set of tools subject to the `--core-tools` allowlist check)".

2. The code's own comment groups `skill` with MCP under "Dynamically discovered tools", a separate exclusion category from the synthetic tools (`agent`, `exit_plan_mode`, …). Lumping `skill` in with the synthetic ones here is functionally fine (both bypass the allowlist) but slightly imprecise vs. the code.

Reviewer: Addressed in b0fe3a7.

1. Rephrased to: "`structured_output` is deliberately excluded from `PermissionManager.CORE_TOOLS` (the set of tools subject to the \`--core-tools\` allowlist check) — alongside the other synthetic tools (\`agent\`, \`exit_plan_mode\`, …)."

2. Added a separate sentence: "Dynamically discovered tools (\`skill\`, MCP) are a separate exclusion category that also bypasses the allowlist for unrelated reasons." — so the two bypass paths are no longer conflated.

`send_message`. The synthetic tool only exists when `--json-schema`
is set; adding it to the `--core-tools` allowlist machinery would mean
`--core-tools read_file --json-schema X` silently drops the terminal
contract.

Explicit `permissions.deny` rules and `--exclude-tools` settings still
apply via `PermissionManager.evaluate` → `isToolEnabled`. If the user
denies the tool, the run hits `maxSessionTurns` and the
`--json-schema` hint in `handleMaxTurnsExceededError` tells them where
to look.

## Subagent contexts

`Config.createToolRegistry` accepts a `forSubAgent: true` option that
suppresses the synthetic registration. Subagent overrides reuse the
parent Config via prototype delegation (`createApprovalModeOverride` /
`buildSubagentContextOverride` → `Object.create(base)`), and
`this.jsonSchema` propagates through the prototype chain. Without the
flag, the synthetic tool would register in the subagent's registry
too, and a subagent calling it would receive the "session ends now"
llmContent — but only `runNonInteractive`'s main / drain loops detect
that as terminal, so the subagent would keep running and burn tokens
on a tool whose contract its loop can't honor.

> 

Author: **[Suggestion]** The subagent tool suppression relies on the single brittle call path through `createToolRegistry(forSubAgent: true)`. Any future subagent mechanism that bypasses this path will leak the synthetic tool into subagents, causing them to loop forever burning tokens. Consider adding a warning for future maintainers or a runtime guard.

```suggestion
Without the flag, the synthetic tool would register in the subagent's registry too, and a subagent calling it would receive the "session ends now" llmContent — but only `runNonInteractive`'s main / drain loops detect that as terminal, so the subagent would keep running and burn tokens on a tool whose contract its loop can't honor.

> **Maintainer note:** Any new subagent mechanism MUST route through `createToolRegistry(forSubAgent: true)` or explicitly filter out the synthetic tool. A runtime guard in `syntheticOutput.ts` that returns `fatalError` when invoked outside the expected context would make this coupling fail-safe rather than fail-silent.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

## MCP shadow-tool guard

`tool-registry.ts:registerTool` checks the lazy `factories` map for
name collisions, not just the eager `tools` map. If an MCP server
discovers a tool literally named `structured_output`, the
auto-qualification path that exists for eager-tool collisions fires
for factory collisions too: the MCP tool gets renamed to
`mcp__<server>__structured_output` and the synthetic factory keeps
the bare name. Without this guard, an MCP server could silently hijack
the structured-output contract.

## Compatibility surface

| Combination | Status | Rationale |
| --- | --- | --- |
| `--json-schema` + `-p` (or stdin, or positional) | Supported | Primary headless path. |
| `--json-schema` + `--output-format text` (default) | Supported | `JSON.stringify(payload)` + newline. |
| `--json-schema` + `--output-format json` / `stream-json` | Supported | `structured_result` field carries the raw object. |

> | `--json-schema` + `--bare` | Supported | Synthetic tool registered alongside the bare three. |

Author: `--bare` is mentioned without definition. The user doc explains it ("registered alongside the bare three: `read_file`, `edit`, `run_shell_command`") — consider a parenthetical here too, or cross-link the user-doc entry. Readers landing on the design doc first won't know what `--bare` means.

Reviewer: Addressed in b0fe3a7. The Permission gating → `--bare` interaction section now explains what bare mode does: "short-circuits the settings → CLI config bridge: builds \`mergedDeny\` as \`[...(bareMode ? [] : settings.permissions.deny), ...]\`, so settings-level denies are dropped under \`--bare\`." Readers landing on the design doc first now have enough context without needing the user doc.

Anthropic. The synthetic-tool approach is provider-agnostic.

**Reorder structured_output to the front of the batch instead of
filtering.** Lets side-effecting siblings run if the structured call
fails validation. Rejected because the contract for `--json-schema` is
"produce structured output" — if the model is in this mode, sibling
side-effects are probably a mistake. Suppressing them entirely is
safer; the model sees a "Skipped:" tool_result and can re-issue them
in a separate turn.

**Local `$ref` resolution inside `schemaRootAcceptsObject`.** Would
catch schemas like `{anyOf: [{$ref: "#/$defs/String"}], $defs: {…}}`
at parse time. Rejected for now because the cost (cycle detection,
JSON Pointer syntax, `$defs` vs `definitions`, partial pointers,
remote refs) outweighs the benefit; the `maxSessionTurns` hint already
points users at "schema is satisfiable" as a likely cause.

## Open work

- Schema-aware response validation could grow a `pattern`-based
  ReDoS guard if real users hit catastrophic-backtracking patterns
  in `--json-schema` arguments.
- SDK protocol additions (Python / TypeScript / Java SDKs exposing a
  typed `structured_result` field) — track separately;

>   [PR #4001](https://github.com/QwenLM/qwen-code/pull/4001) covered

Author: "Covered that scope" is ambiguous about whether #4001 is merged, closed-unmerged, or still open. A one-word status (`merged` / `closed` / `open`) would tell readers what state to expect when they click through.

Reviewer: Addressed in b0fe3a7. Now reads: "(closed unmerged on 2026-05-11) covered that scope before the cli/core work landed and was superseded."

fenced, prefixed with chatter, or hallucinate fields. Tool-call
validation is enforced by the function-calling layer before
`execute()`, which gives us a hard syntactic + semantic guard.

**OpenAI's `response_format: {type: "json_schema", …}`.** Provider-
specific; would require parallel implementations for Gemini and
Anthropic. The synthetic-tool approach is provider-agnostic.

**Reorder structured_output to the front of the batch instead of
filtering.** Lets side-effecting siblings run if the structured call
fails validation. Rejected because the contract for `--json-schema` is
"produce structured output" — if the model is in this mode, sibling
side-effects are probably a mistake. Suppressing them entirely is
safer; the model sees a "Skipped:" tool_result and can re-issue them
in a separate turn.

**Local `$ref` resolution inside `schemaRootAcceptsObject`.** Would
catch schemas like `{anyOf: [{$ref: "#/$defs/String"}], $defs: {…}}`
at parse time. Rejected for now because the cost (cycle detection,
JSON Pointer syntax, `$defs` vs `definitions`, partial pointers,
remote refs) outweighs the benefit; the `maxSessionTurns` hint already

> points users at "schema is satisfiable" as a likely cause.

Author: **[Suggestion]** This inverts the troubleshooting cause: the max-turns hint points users at a schema being **unsatisfiable**, not satisfiable.

```suggestion
points users at "schema is unsatisfiable" as a likely cause.
```

_— gpt-5.5 via Qwen Code /review_

Reviewer: Addressed in 72724d0. Fixed the typo — now reads "schema is **unsatisfiable**" (was "satisfiable").

# Structured Output (`--json-schema`) — Design

This document captures the implementation decisions behind the
`--json-schema` headless feature. User-facing usage lives in
[`docs/users/features/structured-output.md`](../../users/features/structured-output.md).

## Goal

In headless runs (`qwen -p`, piped stdin, or positional prompt), let
the caller constrain the model's final answer to a user-supplied JSON
Schema and surface the validated payload as machine-readable output
that scripts and downstream tooling can consume directly. The model's
incidental prose during planning is allowed, but the run must
terminate with a payload that conforms to the schema, not with
free-form text.

## Approach: synthetic tool whose parameter schema IS the user schema

When `--json-schema` is set, `Config.createToolRegistry` registers a
synthetic `structured_output` tool
([`syntheticOutput.ts`](../../../packages/core/src/tools/syntheticOutput.ts)).
Its `parametersJsonSchema` is exactly the schema the user passed; its
`execute()` returns a stop-message `llmContent`. The function-calling

> API already validates tool args against `parametersJsonSchema`, so

Reviewer: **[Medium]** This phrasing — "The function-calling API already validates tool args against `parametersJsonSchema`" — reads as though the LLM provider performs the validation server-side. In reality, validation is entirely client-side via Ajv in `BaseDeclarativeTool.build()`. The clarification does follow in the next paragraph, but this is the first conceptual statement in the section and is what a skim-reader absorbs.

Suggestion — lead with the client-side detail so there's no room for misreading:

```suggestion
`execute()` returns a stop-message `llmContent`. The tool-call infrastructure
already validates args against `parametersJsonSchema` client-side (via Ajv in
`BaseDeclarativeTool.build()`), so
```

Reviewer: Addressed in the latest push.

"the model returned an answer conforming to the schema" reduces to
"the model successfully called `structured_output`."

Three properties fall out of this for free:

1. **No bespoke validator path.** Ajv-backed `validateToolParams`
   already runs inside `BaseDeclarativeTool.build()` and rejects
   non-conforming args before `execute()` ever fires.
2. **Standard retry behavior.** A validation failure surfaces to the
   model as a tool-call error the same way any other tool's args error
   does. The model sees the Ajv message and can correct in the next
   turn.
3. **Provider-agnostic.** Gemini, OpenAI, and Anthropic all serialize
   tool param schemas the same way; the synthetic tool plugs into all

>    three.

Reviewer: **[Nit]** "all serialize tool param schemas the same way" isn't literally true — Anthropic uses `input_schema`, OpenAI/Gemini use `parameters`. What's provider-agnostic is the `DeclarativeTool` abstraction layer that normalizes the differences. A parenthetical would ground the claim:

```suggestion
3. **Provider-agnostic.** Gemini, OpenAI, and Anthropic all serialize
   tool param schemas the same way (via the `DeclarativeTool`
   abstraction); the synthetic tool plugs into all three.
```

Reviewer: Addressed in the latest push.

| `anyOf`/`oneOf`: no branch admits object | reject |
| `allOf`: any branch is `false` or rejects object | reject |
| Root `$ref` (with or without sibling `type`) | reject |
| `not`: bare `{type: "object"}` (no narrowing keywords) | reject |
| `not`: `{type: "object", required: […], …}` etc. | accept (narrowing keywords leave some objects satisfiable; defer) |
| `if: true` + `then` rejects object | reject |
| `if: false` + `else` rejects object | reject |

**Deferred to Ajv at runtime:**

- `$ref` inside `anyOf` / `oneOf` / `allOf` branches (opaque — local
  `$ref` resolution would need cycle detection, JSON Pointer escapes,
  and `$defs` vs `definitions` handling; the cost outweighs the
  benefit for a parse-time best-effort check).
- `if` whose value is an object schema (decidable only against a
  candidate value).
- Negated `anyOf` / `oneOf` / `const` patterns more complex than
  `not.type`.
- Arbitrary `pattern` ReDoS exposure (user-supplied; the threat model
  is narrow because the flag is a CLI argument, not a network input).

The `maxSessionTurns` exit path appends a `--json-schema`-specific
hint pointing users at the three common stuck-run causes (schema

> unsatisfiable / tool denied / model never called) so the runtime

Reviewer: **[Nit]** "three common stuck-run causes (schema unsatisfiable / tool denied / model never called)" frames all three as co-equal causes. But the actual error message in `errors.ts:264` frames "model never called" as the **symptom**, with "denied" and "unsatisfiable" as the two causes to verify:

> *"If the model never called structured_output, verify it isn't denied by permissions.deny / --exclude-tools and that the schema is satisfiable."*

Consider rephrasing to match the symptom→cause structure:

> "pointing users at the common stuck-run symptom (model never called `structured_output`) and its two likely causes (tool denied via permissions / schema unsatisfiable)"

Reviewer: Addressed in the latest push.

The synthesised "Skipped:" body has two variants:

- **Success path** (a structured call captured the contract this turn):
  `"Skipped: this turn's structured_output contract took precedence as
the terminal output."` — short, because the session terminates
  immediately and no consumer (model or SDK) acts on it.
- **Retry path** (no structured call captured, the model gets another
  turn): adds `"Re-issue this call in a separate turn if needed."` —
  this is the only model-actionable case.

### Main-turn / drain-turn parity

`processToolCallBatch(batchRequests, setModelOverride)` is defined
inside `runNonInteractive` and called from both:

- The main-turn loop (top of the function).
- `drainOneItem` (cron-prompt / background-task notification reply
  loop).

The drain turn matters because `structured_output` is registered for
the whole session, so a cron job or a notification reply MIGHT also
fire the tool. The helper handles both call sites identically;

> the only call-site-specific binding is which `modelOverride` variable

Author: **[Suggestion]** "The helper handles both call sites identically" is accurate for the helper's internal logic, but the post-helper termination control flow differs materially: the main-turn path directly calls `return emitStructuredSuccess()`, while the drain-turn path requires a two-hop termination (`drainOneItem` returns → `drainLocalQueue` detects captured submission → post-drain code calls `emitStructuredSuccess()`). A maintainer debugging a drain-turn-initiated termination would not find this divergence in the design doc.

```suggestion
The helper's internal batch-processing logic is identical for both call sites;
the only call-site-specific binding is which `modelOverride` variable
to write to — passed in as a setter. After the helper returns, callers
diverge: the main-turn loop checks `structuredSubmission` and calls
`emitStructuredSuccess()` directly; the drain-turn path returns from
`drainOneItem`, which causes `drainLocalQueue` to detect the captured
submission, stop draining, and fall through to the post-drain
`emitStructuredSuccess()` call.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

> the only call-site-specific binding is which `modelOverride` variable

Author: **[Suggestion]** "The helper handles both call sites identically" is accurate for the helper internal logic, but the post-helper termination control flow differs: the main-turn path directly calls `return emitStructuredSuccess()`, while the drain-turn path requires a two-hop termination (`drainOneItem` returns → `drainLocalQueue` detects captured submission → post-drain `emitStructuredSuccess()`).

```suggestion
The helper internal batch-processing logic is identical for both call sites;
the only call-site-specific binding is which `modelOverride` variable
to write to — passed in as a setter. After the helper returns, callers
diverge: the main-turn loop checks `structuredSubmission` and calls
`emitStructuredSuccess()` directly; the drain-turn path returns from
`drainOneItem`, which causes `drainLocalQueue` to detect the captured
submission, stop draining, and fall through to the post-drain
`emitStructuredSuccess()` call.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

  zero-arg call against an empty schema lands as `{}` and renders as
  `{}` on stdout, not `null`. The `?? null` step is defence-in-depth
  for the strictly-undefined case.

TEXT mode writes just the `result` field + newline to stdout (any
incidental assistant prose accumulated during the run is discarded —
not mirrored to stderr). JSON mode emits the full event log as a
JSON array; `structured_result` lives on the final `type: "result"`
element of that array, not at the document root. Stream-json mode
emits each message on its own line as JSONL; the terminating `result`
line carries `structured_result`.

## Privacy: cross-surface redaction

The args submitted via `structured_output` ARE the structured payload.
On the success path they're already on stdout; on validation-failure
retries they may never reach stdout at all. Either way, persisting
them on durable on-device surfaces (or exporting them off-device
through telemetry) is duplication that leaks the payload into
longer-lived storage than the user asked for. The redaction rule is
therefore "never persist any args from this synthetic tool, regardless
of outcome," not just "dedup what's already on stdout."

> Two surfaces have to redact, and both share the same placeholder

Author: **[Suggestion]** The "Two surfaces have to redact" enumeration lists telemetry and chat-recording JSONL, but does not mention hooks (`PreToolUse`, `PostToolUse`, `PostToolUseFailure`) as an intentionally non-redacted surface. The user doc has a standalone "Hooks see raw args" blockquote; the design doc should document the same exclusion so future maintainers extending the redaction pipeline know hooks are deliberate.

Suggested addition after the two-surface enumeration:

> Hooks (`PreToolUse`, `PostToolUse`, `PostToolUseFailure`) are intentionally excluded from this redaction — they receive the raw `tool_input` because the hook contract is "see what the tool sees." Audit-style catch-all hooks must filter on `tool_name` or add their own redaction before forwarding payloads off-device.

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

> Two surfaces have to redact, and both share the same placeholder

Author: **[Suggestion]** The "Two surfaces have to redact" enumeration lists telemetry and chat-recording JSONL, but does not mention hooks (`PreToolUse`, `PostToolUse`, `PostToolUseFailure`) as an intentionally non-redacted surface. The user doc has a standalone "Hooks see raw args" blockquote; the design doc should document the same exclusion.

Suggested addition after the two-surface enumeration:

> Hooks (`PreToolUse`, `PostToolUse`, `PostToolUseFailure`) are intentionally excluded from this redaction — they receive the raw `tool_input` because the hook contract is "see what the tool sees."

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

the terminal output."` — short, because the session terminates
  immediately and no consumer (model or SDK) acts on it.
- **Retry path** (no structured call captured, the model gets another
  turn): adds `"Re-issue this call in a separate turn if needed."` —
  this is the only model-actionable case.

### Main-turn / drain-turn parity

`processToolCallBatch(batchRequests, setModelOverride)` is defined
inside `runNonInteractive` and called from both:

- The main-turn loop (top of the function).
- `drainOneItem` (cron-prompt / background-task notification reply
  loop).

The drain turn matters because `structured_output` is registered for
the whole session, so a cron job or a notification reply MIGHT also
fire the tool. The helper handles both call sites identically at
invocation time; the only call-site-specific binding is which
`modelOverride` variable to write to — passed in as a setter.

The **post-helper termination flow** differs between the two sites:
the main-turn path directly calls `return emitStructuredSuccess()`,
while the drain-turn path requires a two-hop termination
(`drainOneItem` returns a sentinel that its caller checks to break out

> of the drain loop before calling `emitStructuredSuccess`). Both

Reviewer: **[Nit — factual inaccuracy]** "`drainOneItem` returns a sentinel" — but `drainOneItem` is `async (): Promise<void>` and returns nothing. The termination detection works via the closure-captured `structuredSubmission` variable, which `processToolCallBatch` mutates as a side effect. The source code comment at `nonInteractiveCli.ts:526` is explicit about this:

> "The helper **mutates the closure-captured** `structuredSubmission` directly"

The "two-hop" architectural description is correct (two levels of callers — `drainLocalQueue` and the holdback loop — each check `structuredSubmission` and exit), but the mechanism isn't a return-value sentinel.

Suggest:

```suggestion
the main-turn path directly calls `return emitStructuredSuccess()`,
while the drain-turn path requires a two-hop termination
(`processToolCallBatch` captures the result into the closure-scoped
`structuredSubmission`; `drainLocalQueue` checks it to stop the drain
loop, then the holdback loop checks it to break out and call
`emitStructuredSuccess`). Both
```

Reviewer: Addressed in the latest push.

### Failure paths

| Cause                                                             | Exit code                     | Surface                                                                                                                                                                                                                                                                            |
| ----------------------------------------------------------------- | ----------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Model emits plain text only                                       | 1                             | Error with turn count + truncated `Output preview`.                                                                                                                                                                                                                                |

> | Model never calls `structured_output` for `maxSessionTurns` turns | 53                            | `Reached max session turns` + `--json-schema` hint pointing at the three common causes.                                                                                                                                                                                            |

Reviewer: **[Nit — internal inconsistency]** This table cell still says "three common causes", but the parse-time validation pipeline section (line ~112–115) was already updated to the symptom → cause framing:

> "common stuck-run symptom (model never called `structured_output`) and its two likely causes (tool denied via permissions / schema unsatisfiable)"

Suggest aligning:

```suggestion
| Model never calls `structured_output` for `maxSessionTurns` turns | 53                            | `Reached max session turns` + `--json-schema` hint pointing at the common stuck-run symptom and its two likely causes.                                                                                                                                                              |
```

Reviewer: Addressed in the latest push.
